include/net/nrf_cloud_codec.h

```diff
@@ -20,37 +29,517 @@ extern "C" {
  * @{
  */
 
+/** @brief Object types for encoding and decoding */
+enum nrf_cloud_obj_type {
+	NRF_CLOUD_OBJ_TYPE__UNDEFINED,
+
+	NRF_CLOUD_OBJ_TYPE_JSON,
+
+	NRF_CLOUD_OBJ_TYPE__LAST,
+};
+
+/** @brief Encoded data sources */
+enum nrf_cloud_enc_src {
+	/** Encoded data is not set */
+	NRF_CLOUD_ENC_SRC_NONE,
+	/** Encoded data was set by @ref nrf_cloud_obj_cloud_encode */
+	NRF_CLOUD_ENC_SRC_CLOUD_ENCODED,
+	/** Encoded data was set manually or by @ref NRF_CLOUD_OBJ_PRE_ENC_DEFINE */
+	NRF_CLOUD_ENC_SRC_PRE_ENCODED,
+};
+
+/** @brief Object used for building nRF Cloud messages. */
+struct nrf_cloud_obj {
+
+	enum nrf_cloud_obj_type type;
+	union {
+		cJSON *json;
+	};
+
+	/** Source of encoded data */
+	enum nrf_cloud_enc_src enc_src;
+	/** Encoded data */
+	struct nrf_cloud_data encoded_data;
+};
+
+/** @brief Define an nRF Cloud JSON object.
+ *
+ * This macro defines a codec object with the type of NRF_CLOUD_OBJ_TYPE_JSON.
+ *
+ * @param _name	Name of the object.
+ */
+#define NRF_CLOUD_OBJ_JSON_DEFINE(_name) \
+	struct nrf_cloud_obj _name = { .type = NRF_CLOUD_OBJ_TYPE_JSON, .json = NULL, \
+				       .enc_src = NRF_CLOUD_ENC_SRC_NONE, \
+				       .encoded_data = { .ptr = NULL, .len = 0 } }
+
+/** @brief Define an nRF Cloud codec object of the specified type.
+ *
+ * @param _name	Name of the object.
+ * @param _type	Type of the object.
+ */
+#define NRF_CLOUD_OBJ_DEFINE(_name, _type) \
+	struct nrf_cloud_obj _name = { 0 }; \
+	_name.type = _type; \
+	_name.enc_src = NRF_CLOUD_ENC_SRC_NONE;
+
+/** @brief Define an nRF Cloud object with pre-encoded data.
+ *
+ * This macro defines a codec object with the type of NRF_CLOUD_OBJ_TYPE__UNDEFINED
+ * and initializes the encoded data with the provided parameters.
+ *
+ * @param _name	Name of the object.
+ * @param _data	Pointer to pre-encoded data.
+ * @param _len	Size of the pre-encoded data.
+ */
+#define NRF_CLOUD_OBJ_PRE_ENC_DEFINE(_name, _data, _len) \
+	struct nrf_cloud_obj _name = { .type = NRF_CLOUD_OBJ_TYPE__UNDEFINED, \
```

Comment:
In order to use this with CoAP, we need to know what the format of the data is (JSON vs. CBOR), so we can set the Content-Type properly.

Reply:
this can be addressed when COAP support is implemented.